src/renderer/components/input/input.tsx

```diff
@@ -106,125 +117,105 @@ export class Input extends React.Component<InputProps, State> {
-        if (!validationId) {
-          this.validationId = validationId = uniqueId("validation_id_");
-        }
-        asyncValidators.push(
-          result.then(
-            () => null, // don't consider any valid result from promise since we interested in errors only
-            error => this.getValidatorError(value, validator) || error
-          )
-        );
-      }
+    try {
+      this.asyncValidating = true;
+
+      return (await Promise.all(
+        this.asyncValidators.map(validator => (
+          validator.validate(value, this.props)
+            .then(isValid => {
+              if (!isValid) {
+                return [this.resolveValidatorMessage(validator.message, value)];
+              }
+
+              return [];
+            })
+            .catch(error => Promise.resolve<React.ReactNode[]>([error, this.resolveValidatorMessage(validator.message, value)]))
+        )),
+      )).flat();
+    } finally {
+      this.asyncValidating = false;
     }
+  }
 
-    // save sync validators result first
-    this.setValidation(errors);
+  @action
+  runValidatorsRaw(value: string) {
+    this.errors = [];
 
-    // handle async validators result
-    if (asyncValidators.length > 0) {
-      this.setState({ validating: true, valid: false });
-      const asyncErrors = await Promise.all(asyncValidators);
+    // run validators
+    for (const validator of this.validators) {
+      const isValid = validator.validate(value, this.props);
 
-      if (this.validationId === validationId) {
-        this.setValidation(errors.concat(...asyncErrors.filter(err => err)));
+      if (!isValid) {
+        if (typeof validator.message === "function") {
+          this.errors.push(validator.message(value, this.props));
+        } else {
+          this.errors.push(validator.message);
+        }
       }
     }
 
-    this.input.setCustomValidity(errors.length ? errors[0].toString() : "");
+    this.inputRef.current.setCustomValidity(this.errors.length ? this.errors[0].toString() : "");
   }
 
-  setValidation(errors: React.ReactNode[]) {
-    this.setState({
-      validating: false,
-      valid: !errors.length,
-      errors,
-    });
-  }
-
-  private getValidatorError(value: string, { message }: InputValidator) {
-    if (isFunction(message)) return message(value, this.props);
-
-    return message || "";
-  }
-
-  private setupValidators() {
-    this.validators = conditionalValidators
-      // add conditional validators if matches input props
-      .filter(validator => validator.condition(this.props))
-      // add custom validators
-      .concat(this.props.validators)
-      // debounce async validators
-      .map(({ debounce, ...validator }) => {
-        if (debounce) validator.validate = debouncePromise(validator.validate, debounce);
-
-        return validator;
-      });
-    // run validation
-    this.validate();
-  }
+  runValidators = debounce(() => this.runValidatorsRaw(this.getValue()), 500, {
```

Comment:
I think some form of denouncing is a good idea. That is because what is considered an "invalid character" depends on the final version.

For instance, a namespace must be a DNS label. Which cannot end in a `-`. I as a user would find it rather annoying to be told that while I am typing unless I pause. Because if I am typing slowly (but within that denounce window) I shouldn't see an error between when I type the `-` and when I type an `a` for instance.

Reply:
Still sounds to me like, for such a type, you should have 2 validators, one to check for invalid characters in real-time while the user is typing (no debouncing), and the other on-submit to check for well-formed input, such as trailing `-`